# CENevo
Scripts and datasets associated with the manuscript "Ancient alleles drive contemporary climate adaptation in an alpine plant"

**Demography**

ABC coalmod speciation data and scripts: Appoximate Bayesian Computation to estimate the speciation time for D. sylvestris and the time to coalescence at the DsCEN/2 locus from a geographic sample of haplotypes

Onset of selection: Appoximate Bayesian Computation workflow to estimate the time of onset of selection for candidate regions

**Distribution modelling**

Scripts to perform haplotype distribution models, with past and future temporal projections

**Ecology**

Data and scripts assessing phenotypic divergence in common gardens, and for evaluating DsCEN/2 allelic effects in transplant experiments

**Genome assembly tools**

Scaffold and gene names used in the manuscript, and workflow to extract annotation for candidate loci

**Genome scans**

LSD scans for detecting loci under selection and characterising genetic trade-offs, as parameterised under an empirically-informed continent-island model

**Genomic offset**

Gene-environment association models based on gradient forests for predicting shifts in genome-wide adaptive composition (i.e. the genomic offset and glacial genomic offset+)

**Haplotype data**

Data and scripts for analyses of CEN/2 and TCF1 haplotypes

**Molecular biology**

Alignment of FT gene family. Transcripts, data and scripts used in Arabidopsis transformation experiment  

**Population genomics**

Population genomic analyses including Poolstat, computation of SFS, Fst genome scans using HMM, computation of thetas, Fst and theta plots for candidate regions, candidate annotation tables, GWAS on flowering time, LD for focal populations and across the species' range, environmental associations and IBD

**Radiation introgression**

ABBA-BABA data and scripts: test introgression across the Dianthus radiation with D-statistics

